Makefile

```diff
@@ -5,10 +5,26 @@ ifeq ($(IDF_PATH),)
 THIS_MK_FILE:=$(notdir $(lastword $(MAKEFILE_LIST)))
 THIS_DIR:=$(abspath $(dir $(lastword $(MAKEFILE_LIST))))
 IDF_PATH=$(THIS_DIR)/sdk/esp32-esp-idf
-all:
-%:
+
+TOOLCHAIN_VERSION:=20181106.0
+PLATFORM:=linux-x86_64
+
+all: toolchain
+%: toolchain
 	@echo Setting IDF_PATH and re-invoking...
-	@env IDF_PATH=$(IDF_PATH) PATH=$(PATH):$(THIS_DIR)/tools/toolchains/esp32/bin/ $(MAKE) -f $(THIS_MK_FILE) $@
+	@env IDF_PATH=$(IDF_PATH) PATH=$(PATH):$(THIS_DIR)/tools/toolchains/esp32-$(PLATFORM)-$(TOOLCHAIN_VERSION)/bin $(MAKE) -f $(THIS_MK_FILE) $@
+
+toolchain: $(THIS_DIR)/tools/toolchains/esp32-$(PLATFORM)-$(TOOLCHAIN_VERSION)/bin/xtensa-esp32-elf-gcc
+	@echo Checking pre-compiled toolchain
```

Comment:
I'd suggest not using a phony `toolchain` target, as that will lead to unnecessarily rebuilding targets not really in need of rebuilding (which might be what broke Travis, since it got stuck in a menuconfig). What I'd do instead is:
```
ESP32_GCC:=$(THIS_DIR)/tools/toolchains/esp32-$(PLATFORM)-$(TOOLCHAIN_VERSION)/bin/xtensa-esp32-elf-gcc
ESP32_TOOLCHAIN_DL:=$(THIS_DIR)/cache/toolchain-esp32-$(PLATFORM)-$(TOOLCHAIN_VERSION).tar.xz

all: | $(ESP32_GCC)
%: | $(ESP32_GCC)
    ...

$(ESP32_GCC): $(ESP32_TOOLCHAIN_DL)
    ...
```
Note the pipes (`|`) on the $(ESP32_GCC) deps; If you're not familiar with them,  they turn the following dependencies into "ordering only" dependencies, so dependent targets won't get marked for rebuild if the time-stamp on the dependency is newer than the target in question. Here, that's effectively "we need a toolchain before we can build, but we don't care about the actual timestamp on the gcc binary, only that it exists".

Reply:
Thanks for the hint.